protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftServiceManager.java

```diff
@@ -69,6 +71,7 @@
   private final ThreadContextFactory threadContextFactory;
   private final RaftLog log;
   private final RaftLogReader reader;
+  private final Map<Long, CompletableFuture> futures = Maps.newHashMap();
```

Comment:
@kuujo shouldn't this be a concurrent hash map?

Reply:
The state machine is single threaded since changes have to be applied in order